Stronger chain detection in LoopCarry pass

vksnk opened the pull request:
```can_prove``` is stronger than ```graph_equal```, because it doesn't require index expressions to be *exactly* the same, but evalutate to the same value. I kept the ``` graph_equal``` check, because it's faster and should be executed before the more expensive check.

In one of the internal workloads, I see that with this change, what was previously split into three different chains of 4-, 2-, 3- values, is correctly combined into one long chain of lenght 9-.

steven-johnson commented on src/LoopCarry.cpp:
```diff
@@ -301,9 +301,9 @@ class LoopCarryOverLoop : public IRMutator {
                 if (loads[i][0]->name == loads[j][0]->name &&
                     next_indices[j].defined() &&
-                    graph_equal(indices[i], next_indices[j]) &&
+                    (graph_equal(indices[i], next_indices[j]) || can_prove(indices[i] == next_indices[j])) &&
```
Please add the comment from this PR here and below for future readers

vksnk commented:
Also, fixed a bug when indices with different types are compared.

BTW, as far as I know, this pass is only used in Hexagon and Xtensa backends.

vksnk commented:
All tests are green now.

abadams commented:
See the comment up at line 250. It's not safe to use can_prove on a boolean Expr after doing substitute_in_all_lets. To make it safe to call, you have to call common_subexpression_elimination on the Expr first.

Note that this gets called on every pair of indices, so it has quadratic complexity in the IR size. I worry that this will stall for very large unrolled stencils. It's worth writing a test of a very large case. If it does indeed stall, we might need a better algorithm. One could for example hash the expressions and look for hash collisions, where by "hash" I mean substitute in some arbitrary values for the variables and constant-fold.

abadams commented on src/LoopCarry.cpp:
```diff
@@ -299,11 +299,16 @@ class LoopCarryOverLoop : public IRMutator {
                     continue;
                 }
+                // can_prove is stronger than graph_equal, because it doesn't require index expressions to be
+                // exactly the same, but evalutate to the same value. We keep the graph_equal check, because
```
evalutate -> evaluate

vksnk commented:
> See the comment up at line 250. It's not safe to use can_prove on a boolean Expr after doing substitute_in_all_lets. To make it safe to call, you have to call common_subexpression_elimination on the Expr first.
> 
> Note that this gets called on every pair of indices, so it has quadratic complexity in the IR size. I worry that this will stall for very large unrolled stencils. It's worth writing a test of a very large case. If it does indeed stall, we might need a better algorithm. One could for example hash the expressions and look for hash collisions, where by "hash" I mean substitute in some arbitrary values for the variables and constant-fold, and then only do can_prove on exprs that have the same hash.

Thanks a lot, this is very helpful! 

I changed it to apply CSE first and only then run can_prove. Also, added a test which triggers loop_carry on the loop with large number of indices and the compilation time seems to be fine.

vksnk commented on test/correctness/loop_carry.cpp:
```diff
@@ -0,0 +1,69 @@
+#include "Halide.h"
+#include <stdio.h>
+
+// This file demonstrates two example custom lowering passes. The
```
Fixed

vksnk commented:
I don't think test failures are related.
